### ISSUE TITLE
Report of Reviewer 2

### ISSUE THREAD

**Issue by NelleV** [computorg/published-202301-chagneux-macrolitter#2]
Associate Editor: Nelle Varoquaux
Reviewer 2:

# Reviewer 2: Reviewing history

- Paper submitted March 3, 2022
- Reviewer invited July 1st 2022
- Review 1 received August, 11th, 2022
- Paper revised October 3rd, 2022
- Reviewer invited October 17th, 2022
- Review 2 received December 1st, 2022

# Round 1

This paper presents a method to count macrolitter items from video footages of riverbanks. The task considered is therefore an interesting scientific challenge, as it involves Multiple Objects Tracking on unusually moving videos (boat-embedded cameras), and a useful application of computer vision, as it can help measure the amount of waste along watercourses, and relate this to other environmental data. Furthermore, the presented method relies on statistical tools (machine learning, Kalman filtering), which fall within the scope of Computo journal.

Overall, I find the paper well written, the proposed method elegant and clearly exposed, and well situated with respect to the state of the art. I believe the method is worth publishing, but some concerns need to be addressed before I could recommend the publication of the manuscript.

My major concern regards the evaluation of the method, and especially its comparison to other algorithms. As will be detailed below, I have a doubt on the fairness of this comparison, as I understand that some parameters of the proposed method were tuned to optimise a metric on the test set (which would already be questionable), whereas other methods do not benefit from such optimization (would they, at least the comparison would be fair, although the performance estimation would be over-optimistic). I hope this is a misunderstanding and that the authors can clarify this point.

> **Authors' response** Regarding the main concern on the evaluation methodology, we have modified our
protocol which was, indeed, not sufficiently rigorous. As is now explained in the "Experiments" section (and in Appendix), we computed for the 3 methods separately the set of hyperparameters $\tau$ and $\kappa$ on a distinct split of the
evaluation footage (which was divided in 2). We then evaluated the methods using their respective optimal points on the other split of the data. All results were updated accordingly in the revised version

Besides (but this may not be the authors’ fault) I was suprised to only have a pdf file to review: no data was available, either on the reviewing platform or through a link shared in the manuscript; no code was shared, except the code chunks printed in the pdf file. This is a problem when reviewing a paper submitted to a journal where “The reproducibility of numerical results is a necessary condition for publication” and where " submissions must include all necessary data
(e.g. via zenodo repositories) and code."

> **Editor's response** Due to a mistake on my part, the link to the github repository was not send to the reviewers.

>**Authors' response** Concerning the reproducibility of the work and the availability of the code chunks, we point to the remarks of the Editorial Board which has apparently ommitted to share our implementation. We hope that the repository used to generate the paper will be sufficient material to clarify the reproducibility of our work.

Finally, but this is a minor point and only a suggestion, I think the authors could improve the reproducibility of the method and its evaluation by detailing some crucial points (equations, algorithms) instead of referring to other papers. For example, referring to the Hungarian method is fine, but avoiding the formal definitions of some metrics by referring to the paper where they were initially defined, is not, in my opinion. Also, I think Kalman filter updates in the
present case is worth detailing.

Below, you will find more detailed comments following the structure of the paper.

## Introduction

Reference to code and data so that code chunks can be run?

### Training

- “such that $Y^{xy}$ is the probability that $(x, y)$ is the center of an object” -> such that $(px, py)$ is the center of an object in $I$

### Approximations of the filtering distributions

 - In the EKF approximation: the $\nabla$ symbol may be replaced by the differential symbol $\partial$, as the optical flow is not a scalar function. Furthermore, how is this differential estimated?
- “This allows the implementation of Kalman updates.” -> Although the Kalman filter is a well established method, you may give more precisions about the updates in this very case, using your notations; it would improve the reproducibility of your method.

### Counting
-  Why not giving the equation of the moving average, to avoid any ambiguity?
- choice of parameter $\nu$?
- “(see here for a more comprehensive study)” -> no link provided
> **Authors' response**  We have properly defined the moving average $\mathcal{M}_{\ell}^{\kappa}$ used in the postprocessing step.

----
### Detection

- I don’t think referring to equations of another paper is a good practice to define formally metrics that will be used to evaluate the proposed method ; I suggest to report the equations in the paper. Furthermore, in the present case, I have not found equations 23, 24, 26 and 27, neither in reference [12], nor in [41], nor in [42].
- “In brief, a low AssPr implies that several objects are often mingled into only one track, resulting in undercount. A low AssRe implies that single objects are often associated with multiple tracks. If no method is used to discard redundant tracks this results in overcount.” -> It seems that Pr and Re were exchanged here, as undercount usually results in low recall and overcount in low precision.
- “AssRe does not account for tracks predicted from streams of false positive detections generated by the detector (e.g. arising from rocks, water reflections, etc)” -> Isn’t this the captured in AssPr?

> **Authors' response** Considering the association metrics, we apologize for a mistake in the bibliography: [12] pointed to the wrong paper where the equations were not present. In any case, we have now written the defining equations for AssPr and AssRe directly in the body of the paper, as requested. Note that AssPr and AssRe have *not* been exchanged. The understanding of these quantities is less intuitive than traditional detection recall and precision. Quoting directly from the original paper of the HOTA metrics: "E.g. a low AssRe will result when a tracker splits an object up into multiple predicted tracks. Association precision measures how well predicted trajectories keep to tracking the same ground-truth trajectories. E.g. a low AssPr will result if a predicted track extends over multiple objects.". In a counting system such as ours, a low AssRe reveals fragmented tracks, causing redundant counts and therefore overcounting, and a low AssPr reveals that multiple objects followed by a single track, and therefore undercounting. Note that these metrics are computed at a given detection performance, and therefore it is possible to reach a maximal values AssPr/AssRe for even if if some objects are missed entirely

### Count decomposition

- “At evaluation, we assign each predicted track to either none or at most one ground truth track” -> How so? What is the criterion to assign j to i or to none?
- “2. … the number of redundant counts per ground truth object.” -> Isn’t it the absolute number of redundant counts?
- Why not defining precision and recall based on $N$, $N_{true}$ and $N_{false}$?

> **Authors' response** : Your question on defining a count recall and precision is completely relevant: we have now defined two quantities and which fully leverage our count error decomposition, and we use these in the table for the results. These metrics complement well the graph on the decomposition of the errors and make the results less redundant. We warmly thank reviewer 2 for their advice as these metrics clearly highlight the benefits of our solution over alternatives on the main task of interest. We also added details on how associations are performed to build these metrics.

### Statistics

- “Splitting the original sequences into shorter independent sequences of equal length allows to compute basic statistics.” -> What is the equal length of these sequences?

---

## Experiments

 - “object positions.Similar to our system” -> Missing space

### Detection

 - Please precise that Pr and Det are given as percentages (all the more so as their definitions are not provided in the text)

### Counts

- "We run our algorithm with $\kappa = 7$, $\tau = 5$ , values obtained after a simple hyperparameter study described in Details on the image dataset.” -> In that hyperparameter study, a few important points are unclear:
1.  “ $\kappa$ and $\tau$ must be chosen jointly to decrease as much as possible without increasing $N_{miss}$” -> if I understand correctly, these parameters, which are part of the proposed method, are tuned to optimize the metric on the evaluation (ot test) data (as videos with annotated tracks were not used for training). I hope this is a misunderstanding, otherwise this would be a problem in the evaluation of the method, especially for the comparison with other algorithms, which do not benefit from such an optimization. 
2. What do the colours stand for in the three plots of “Ns” as functions of $\tau$? From the code it seems to correspond to several versions of the algorithm, but please add a legend and explain in the text. Also, a minor point: the labels of the y axis are not consistent with the notations used in the text (N_missing instead of N_miss, N_incorrect instead of N_false, N_redundant instead of N_red)
3. Again, for the three tables of p. 17, please add a legend. It is not desirable that the reader should parse the code in order to understand the plots and tables. Furthermore, this presentation makes it hard to compare the three methods; I’d rather have one table for each sequence $S_i$ (and one for “All” if you like), and one line for each method, inside each table.

> **Authors' response**: Regarding the main concern on the evaluation methodology, we have modified our protocol which was, indeed, not sufficiently rigorous. As is now explained in the Answer to reviews2"Experiments" section (and in Appendix), we computed for the 3 methods separately the set of hyperparameters $\tau$ and $\kappa$ on a distinct split of the evaluation footage (which was divided in 2). We then evaluated the methods using their respective optimal points on the other split of the data. All results were updated accordingly in the revised version.

# Round 2

The authors have seriously taken into account my suggestions and concerns, and I believe the manuscript is now worth publishing.

Here below are some additional/remaining suggestions for the final version.

  - In Approximations of the filtering distributions, all I am saying about the gradient symbol, is that the gradient is a vector
    (e.g. in $R^n$) and it is defined for real valued functions (e.g. $f : R^n -> R$).     The optical flow is not real valued as it maps $R^2$ to $R^2$.

> **Authors' response**  In the section related to the optical flow, we’ve replaced the symbol $\nabla$ with a $\partial$ to refer to the general derivative acting on the multivariate-multivalued optical flow operator. 

- In Computing the confidence regions, I still think there is a typo, and k should be n. Furthermore, now the equation is partially cut out.

- Still in Computing the confidence regions, a word on the estimations     of $E[Z_k^l | Z_{1:k-1}^l]$ and $V[Z_k^l | Z_{1:k-1}^l]$ could be useful, as the Bayesian filtering gives these moments for $X_k$ given $Z_{1:k}$, not $Z_k$ given $Z_{1:k}$.

> **Authors' response** We moved the section “Computing the confidence regions” under the section “Bayesian filtering” to provide more derivations regarding the distribution of  $Z_n$  given $Z_{1:n-1}$ from the filtering distributions $X_{n-1}$ given $Z_{1:n-1}$ . 

- In Count decomposition, the choice to average the results over the different values of the threshold \alpha, is not obvious. Maybe the authors should explain this choice, precising in particular that the performance does not necessarily increase with \$alpha$.

> **Authors' response** In “Count metrics” we’ve added some complements regarding the computation of
the aggregate results over different location thresholds with a small justification for
why we choose the average

-  In Statistics, a closing parenthesis should be removed (typo)
